x-pack/legacy/plugins/apm/public/projections/util/merge_projection.ts

```diff
@@ -0,0 +1,32 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+import { merge, cloneDeep, isPlainObject } from 'lodash';
+import { Projection } from '../typings';
+
+type PlainObject = Record<string | number | symbol, any>;
+
+type DeepMerge<T, U> = U extends PlainObject
+  ? (T extends PlainObject
+      ? (Omit<T, keyof U> &
+          {
+            [key in keyof U]: T extends { [k in key]: any }
+              ? DeepMerge<T[key], U[key]>
+              : U[key];
+          })
+      : U)
+  : U;
+
+export function mergeProjection<T extends Projection, U>(
+  target: T,
+  source: U
+): DeepMerge<T, U> {
+  return merge(cloneDeep(target), source, (a, b) => {
+    if (isPlainObject(a) && isPlainObject(b)) {
+      return undefined;
+    }
+    return b;
+  }) as DeepMerge<T, U>;
+}
```

Comment:
What's the difference between `mergeProjection` and:
```js
_.merge({}, a, b);
```

mostly I don't understand what the last argument in `merge` is for.

Reply:
`_.merge`'s default merge behaviour when it comes to arrays is pretty wild: https://codesandbox.io/s/dawn-fast-um27n. That's an issue when merging clauses/filters. `mergeProjection` basically only merges plain objects, everything else is just overridden.

Reply:
```js
_.merge({}, {must: [{ term: "a" }]}, { must: [{ terms: "b" }] });
```

returns:
```json
{
  "must": [ { "term": "a", "terms": "b"} ]
}
```

```js
mergeProjection({must: [{ term: "a" }]}, { must: [{ terms: "b" }] })
```

returns:
```json
{
  "must": [ { "terms": "b"} ]
}
```

Is that intentional that it overwrite the first? There should probably be a test to document this behaviour.

Reply:
Even if we want this we can still skip the `cloneObject` right? 
```js
merge({}, a, b)
```

Reply:
Yeah, we can skip cloneDeep, not sure why I added it there. Maybe I don't trust `merge` anymore 🤷‍♂ . I've added a test for `mergeProjection` as well.